components/html-block/html-block.js

```diff
@@ -0,0 +1,271 @@
+				const startup = window.MathJax.startup;
+
+				//
+				//  Extend HTMLAdaptor to handle shadowDOM as the document
+				//
+				class ShadowAdaptor extends HTMLAdaptor {
+					body(doc) {
+						return doc.body || (doc.firstChild || {}).lastChild || doc;
+					}
+					create(kind, ns) {
+						const document = (this.document.createElement ? this.document : this.window.document);
+						return (ns ?
+							document.createElementNS(ns, kind) :
+							document.createElement(kind));
+					}
+					head(doc) {
+						return doc.head || (doc.firstChild || {}).firstChild || doc;
+					}
+					root(doc) {
+						return doc.documentElement || doc.firstChild || doc;
+					}
+					text(text) {
+						const document = (this.document.createTextNode ? this.document : this.window.document);
+						return document.createTextNode(text);
+					}
+				}
+
+				//
+				//  Extend HTMLHandler to handle shadowDOM as document
+				//
+				class ShadowHandler extends HTMLHandler {
+					create(document, options) {
+						const adaptor = this.adaptor;
+						if (typeof(document) === 'string') {
+							document = adaptor.parse(document, 'text/html');
+						} else if ((document instanceof adaptor.window.HTMLElement || document instanceof adaptor.window.DocumentFragment) && !(document instanceof window.ShadowRoot)) {
+							const child = document;
+							document = adaptor.parse('', 'text/html');
+							adaptor.append(adaptor.body(document), child);
+						}
+						//
+						//  We can't use super.create() here, since that doesn't
+						//  handle shadowDOM correctly, so call HTMLHandler's parent class
+						//  directly instead.
+						//
+						return AbstractHandler.create.call(this, document, options);
+					}
+				}
+
+				//
+				//  Register the new handler and adaptor
+				//
+				startup.registerConstructor('HTMLHandler', ShadowHandler);
+				startup.registerConstructor('browserAdaptor', () => new ShadowAdaptor(window));
+
+				//
+				//  A service function that creates a new MathDocument from the
+				//  shadow root with the configured input and output jax, and then
+				//  renders the document.  The MathDocument is returned in case
+				//  you need to rerender the shadowRoot later.
+				//
+				window.MathJax.typesetShadow = function(root) {
+					const InputJax = startup.getInputJax();
+					const OutputJax = startup.getOutputJax();
+					const html = mathjax.document(root, { InputJax, OutputJax });
+					html.render();
+					return html;
+				};
+
+				//
+				//  Now do the usual startup now that the extensions are in place
+				//
+				window.MathJax.startup.defaultReady();
+			}
+		}
+	};
+
+	return new Promise(resolve => {
+		if (!document.head.querySelector('#d2l-mathjax')) {
+			const script = document.createElement('script');
+			script.id = 'd2l-mathjax';
+			script.async = 'async';
+			script.onload = resolve;
+			script.src = 'https://s.brightspace.com/lib/mathjax/3.1.2/mml-chtml.js';
+			document.head.appendChild(script);
+		}
```

Comment:
If Mathjax is already loaded this Promise won't resolve... should you call `resolve()` in an `else` block?

Reply:
That's a good point.  I think I might adjust this to be more like:
``` javascript
let mathJaxLoaded;
const loadMathJax = () => {
    if (mathJaxLoaded) return mathJaxLoaded;
    ...
    mathJaxLoaded = new Promise(resolve => {
        // add the script and resolve
    });
    return mathJaxLoaded;
};
```
Then the consumer (currently just `d2l-html-block`) doesn't need to worry about making a conditional check, it can just:
```
await loadMathJax();
```
But then maybe `requireMathJax` would be a better name.